[FEATURE] ABCDiag - Ne pas afficher la landing page pour les campagnes “ForAbsoluteNovice” (PIX-2129).

alexandrecoin opened the pull request:
## :unicorn: Problème

Dans les campagnes destinées à un public débutant, nous ne souhaitons pas afficher la page d'accueil de la campagne.

## :robot: Solution

Création d'une nouvelle condition d'entrée sur la campagne sur `mon-pix/app/routes/campaigns/start-or-resume.js` qui reprend la création d'une `campaignParticipation` telle que réalisée dans la méthode `_shouldStartCampaignParticipation`, et qui se base sur la propriété `isCampaignForNoviceUser`  (Boolean) de la campagne désirée.

## :rainbow: Remarques

Le fichier `campaigns/start-or-resume.js` agit comme un hub central au routage des campagnes et commence à dégager une très grande complexité, faite de beaucoup de conditions et de redirections, parfois cycliques vers ce même fichier, créant un couplage fort et des difficultés à pouvoir tester le code de façon simple.

## :100: Pour tester

Dans la BDD de la RA, setter `isCampaignForNoviceUser` à `true` pour la campagne `AZERTY123`
Se connecter à une campagne via le code `AZERTY123`
Ne pas voir la landing page avant d'arriver sur l'assessment.

alexandrecoin committed:
Redirect to assessment page without landing page

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2577.review.pix.fr
- Orga: https://orga-pr2577.review.pix.fr
- Certif: https://certif-pr2577.review.pix.fr
- Admin: https://admin-pr2577.review.pix.fr
- API: https://api-pr2577.review.pix.fr/api/

 Please check it out!

alexandrecoin committed:
bsr: reorder start-or-resume class and methods

alexandrecoin committed:
Remove return from start campaign action

alexandrecoin committed:
Check hasSeenLandingPage and isForNovice

We redesigned the logic of the previous commit.
Instead of checking the state and redirect in beforeModel, we now use the original design of the start-or-resume file.
We skip the landing page for novice users in the redirect life cycle method.

MelanieMEB commented on mon-pix/app/routes/campaigns/start-or-resume.js:
```diff
@@ -199,6 +217,10 @@ export default class StartOrResumeRoute extends Route.extend(SecuredRouteMixin)
     return !this.state.doesUserHaveOngoingParticipation;
   }
 
+  get _shouldNotVisitLandingPageAsNovice() {
+    return this.state.isCampaignForNoviceUser;
+  }
+
```
Cette méthode n'est plus utile

MelanieMEB commented on mon-pix/app/routes/campaigns/start-or-resume.js:
```diff
@@ -16,98 +16,19 @@ export default class StartOrResumeRoute extends Route.extend(SecuredRouteMixin)
+
     this._updateStateFrom({ campaign, queryParams: transition.to.queryParams, session: this.session });
 
+    if (this.state.isCampaignForNoviceUser) {
+      this.state.hasUserSeenLandingPage = true;
+    }
```
**proposition:**
Déplacer cette logique dans `this._updateStateFrom`

A la place de : 
```
    const hasUserSeenLandingPage = this._handleQueryParamBoolean(queryParams.hasUserSeenLandingPage, this.state.hasUserSeenLandingPage);
```
mettre
```
const isCampaignForNoviceUser = get(campaign, 'isForAbsoluteNovice', this.state.isCampaignForNoviceUser)
    const hasUserSeenLandingPage = isCampaignForNoviceUser ? true : this._handleQueryParamBoolean(queryParams.hasUserSeenLandingPage, this.state.hasUserSeenLandingPage);

```

alexandrecoin replied:
Bien vu !

celineung replied:
Je reprends la réponse de @MelanieMEB, mais je suggère de ne pas modifier `hasUserSeenLandingPage` (car le nom de la variable ne serait plus correct) mais plutôt de rajouter une variable dans le state:

```
const hasUserNotSeenLandingPage = queryParams.hasUserSeenLandingPage == null ? null : !queryParams.hasUserSeenLandingPage;
const shouldDisplayLandingPage = isCampaignForNoviceUser ? false : this._handleQueryParamBoolean(hasUserNotSeenLandingPage, this.state.shouldDisplayLandingPage);

this.state = {
...
  shouldDisplayLandingPage,
...
}
```

Dans le `resetState`:
```
this.state = {
  shouldDisplayLandingPage: true;
}
```

Cela permet ensuite de **supprimer** `hasUserSeenLandingPage` du state. Et partout où on a besoin de `hasUserSeenLandingPage`, on utilise `shouldDisplayLandingPage` à la place.

alexandrecoin replied:
Entièrement ok pour ce changement, mais on est d'accord pour initialiser `shouldDisplayLandingPage` à `true` de base dans le state plutôt que `false` non ?
Qui du coup inversera tous les appels à l'ancien `hasUserSeenLandingPage ` dans les conditions.

celineung replied:
Oui, l'initialiser à `false`. Je modifie ma réponse précédente.

alexandrecoin committed:
Replace hasUserSeenLandingPage property by shoulDisplayLandingPage
